File: d11_monkey_in_the_middle/src/main.rs
use std::{fs, mem, str::FromStr};

struct Monkey {
    items: Vec<i64>,
    operation: Box<dyn Fn(i64) -> i64>,
    test: [i64; 3],
}

impl Monkey {
    fn get_operation(i: usize) -> Box<dyn Fn(i64) -> i64> {
        let operations = [
            |x| x * 7,
            |x| x + 4,
            |x| x + 2,
            |x| x + 7,
            |x| x * 17,
            |x| x + 8,
            |x| x + 6,
            |x| x * x,
        ];
        // let operations = [|x| x * 19, |x| x + 6, |x| x * x, |x| x + 3];
        Box::new(operations[i])
    }
}

impl FromStr for Monkey {
    type Err = ();
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut s = s.split('\n');
        let i = s
            .next()
            .unwrap()
            .chars()
            .nth(7)
            .unwrap()
            .to_digit(10)
            .unwrap()
            .try_into()
            .unwrap();

        let items = s
            .next()
            .unwrap()
            .split(": ")
            .nth(1)
            .unwrap()
            .split(", ")
            .map(|s| s.parse::<i64>().unwrap())
            .collect();

        s.next().unwrap();
        let operation = Self::get_operation(i);

        let test = s
            .map(|line| line.split(' ').last().unwrap().parse::<i64>().unwrap())
            .collect::<Vec<_>>()
            .try_into()
            .unwrap();

        Ok(Self {
            items,
            operation,
            test,
        })
    }
}

fn main() {
    let path = "input";
    // let path = "example";
    let buf = fs::read_to_string(path).unwrap();

    let mut monkeys: Vec<_> = buf
        .trim()
        .split("\n\n")
        .map(|block| block.parse::<Monkey>().unwrap())
        .collect();

    let lcm = monkeys
        .iter()
        .map(|monkey| monkey.test[0])
        .reduce(|acc, x| acc * x)
        .unwrap();

    print_monkeys(&monkeys);
    println!("lcm: {}", lcm);

    let mut inspections = vec![0; monkeys.len()];

    for _round in 0..10000 {
        for i in 0..monkeys.len() {
            let items = mem::take(&mut monkeys[i].items);
            inspections[i] += items.len();
            for mut item in items {
                let monkey = &mut monkeys[i];
                item = (monkey.operation)(item) % lcm;

                let divider = monkey.test[0];
                let true_throw = monkey.test[1] as usize;
                let false_throw = monkey.test[2] as usize;

                if item % divider == 0 {
                    monkeys[true_throw].items.push(item);
                } else {
                    monkeys[false_throw].items.push(item);
                }
            }
        }
        // print_monkeys(&monkeys);
    }

    print_monkeys(&monkeys);
    println!("{:?}", inspections);
    inspections.sort();
    let value = inspections
        .into_iter()
        .rev()
        .take(2)
        .reduce(|acc, x| acc * x)
        .unwrap();
    println!("{}", value)
}

fn print_monkeys(monkeys: &[Monkey]) {
    let monkeys = monkeys.iter().map(|m| &m.items).collect::<Vec<_>>();
    for monkey in monkeys {
        println!("{:?}", monkey);
    }

    println!();
}

File: d11_monkey_in_the_middle/src/main.py
from typing import Callable, List
from pathlib import Path

operations = [
    lambda x: x * 19,
    lambda x: x + 6,
    lambda x: x * x,
    lambda x: x + 3,
]


class Monkey:
    __slots__ = ("items", "operation", "test")

    def __init__(
        self, items: List[int], operation: Callable[[int], int], test: List[int]
    ) -> None:
        self.items = items
        self.operation = operation
        self.test = test

    def __repr__(self) -> str:
        # return f"Monkey({self.items!r}, {self.operation!r}, {self.test!r})"
        return f"Monkey({self.items})"


def read_monkeys() -> List[Monkey]:
    here = Path(__file__).parent
    path = here.parent / "example"
    monkeys = []

    with open(path) as f:
        f = f.read()
        for i, block in enumerate(f.split("\n\n")):
            block = block.split("\n")
            items = [int(n) for n in block[1].split(": ")[1].split(", ")]
            operation = operations[i]
            test = [int(line.split(" ")[-1]) for line in block[3:]]
            monkeys.append(Monkey(items, operation, test))

    return monkeys


def main() -> None:
    monkeys = read_monkeys()

    print("\n".join(str(m) for m in monkeys))
    print()

    inspection_count = [0] * len(monkeys)

    for i in range(20):
        if i % 10 == 0:
            print(f"round {i}")

        for i, monkey in enumerate(monkeys):
            for item in monkey.items:
                inspection_count[i] += 1
                item = monkey.operation(item) // 3

                divider = monkey.test[0]
                true_throw = monkey.test[1]
                false_throw = monkey.test[2]

                if item % divider == 0:
                    monkeys[true_throw].items.append(item)
                else:
                    monkeys[false_throw].items.append(item)

            monkey.items = []

    print("\n".join(str(m) for m in monkeys))
    print(f"{inspection_count=}")

    value1, value2 = sorted(inspection_count)[-2:]
    print(f"part1 answer: {value1 * value2}")


if __name__ == "__main__":
    main()
